### PR TITLE
Update `checkStakeKeyRegistered` to use `foldEpochState`

### PR DESCRIPTION
**Pull request by Jimbo4350**
# Description

Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
  - integration tests
  See [Runnings tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-testnet/src/Testnet/Components/SPO.hs [hunk 1]
```diff
@@ -75,37 +86,75 @@ checkStakePoolRegistered tempAbsP execConfig poolColdVkeyFp outputFp =
...
+        (DelegationsAndRewards (rewardsMap, _delegMap))
+          <- H.noteShowM $ H.jsonErrorFail $ Aeson.fromJSON @DelegationsAndRewards pledgerStakeInfo
+        H.failWithCustom GHC.callStack Nothing
+                       $ unlines [ "Stake address in question: "
+                                 , Text.unpack (serialiseToBech32 sAddr)
+                                 , "was not registered"
+                                 , "Current stake info for address in question: "
+                                 , show $ map serialiseToBech32 $ Map.keys rewardsMap
+                                 , "foldEpochStateError: " <> show e
+                                 ]
```
Comment (smelc) [inline cardano-testnet/src/Testnet/Components/SPO.hs:130]: Providing a detailed error message, nice :+1:

**Review by smelc — APPROVED** [review]
I'm not familiar with the part of the ledger's state being accessed, but this LGTM

File: cardano-testnet/src/Testnet/Components/SPO.hs [hunk 2]
```diff
@@ -75,37 +86,75 @@ checkStakePoolRegistered tempAbsP execConfig poolColdVkeyFp outputFp =
...
+      Left e -> do
+        void $ execCli' execConfig
+          [ "query", "stake-address-info"
+          , "--address", stakeAddr
+          , "--out-file", oFpAbs
+          ]
+
+        pledgerStakeInfo <- H.leftFailM $ H.readJsonFile oFpAbs
+        (DelegationsAndRewards (rewardsMap, _delegMap))
+          <- H.noteShowM $ H.jsonErrorFail $ Aeson.fromJSON @DelegationsAndRewards pledgerStakeInfo
```
Comment (carbolymer) [inline cardano-testnet/src/Testnet/Components/SPO.hs:122]: can be simplified a bit
```suggestion
        (DelegationsAndRewards (rewardsMap, _delegMap)) <- H.noteShowM $ H.readJsonFileOk oFpAbs
```

File: cardano-testnet/src/Testnet/Components/SPO.hs [hunk 3]
```diff
@@ -310,14 +362,17 @@ registerSingleSpo identifier tap@(TmpAbsolutePath tempAbsPath') cTestnetOptions
            ]
   -- TODO: Currently we can't propagate the error message thrown by checkStakeKeyRegistered when using byDurationM
   -- Instead we wait 15 seconds
```
Comment (carbolymer) [inline cardano-testnet/src/Testnet/Components/SPO.hs:364]: This TODO is obsolete now
```suggestion
  -- TODO: Currently we can't propagate the error message thrown by checkStakeKeyRegistered when using byDurationM
  -- Instead we wait 15 seconds
```

**Review by carbolymer — APPROVED** [review]

**Comment by carbolymer** [comment]
There are build failures in merge queue https://github.com/IntersectMBO/cardano-node/actions/runs/8605681309/job/23582460168#step:13:594

**Pull request merged**